src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp

```diff
@@ -710,6 +712,54 @@ void ShenandoahHeap::decrease_committed(size_t bytes) {
   _committed -= bytes;
 }
 
+// For tracking usage based on allocations, it should be the case that:
+// * The sum of regions::used == heap::used
+// * The sum of a generation's regions::used == generation::used
+// * The sum of a generation's humongous regions::free == generation::humongous_waste
+// These invariants are checked by the verifier on safepoints.
+//
+// Additional notes:
+// * When a mutator's allocation request causes a region to be retired, the
+//   free memory left in that region is considered waste. It does not contribute
+//   to the usage, but it _does_ contribute to allocation rate.
+// * The bottom of a PLAB must be aligned on card size. In some cases this will
+//   require padding in front of the PLAB (a filler object). Because this padding
+//   is included in the region's used memory we include the padding in the accounting.
+// * Mutator allocations are used to compute an allocation rate. They are also
+//   sent to the Pacer for those purposes.
+// * There are three sources of waste:
+//  * The padding used to align a PLAB on card size
+//  * Region's free is less than minimum TLAB size and is retired
```

Comment:
I think there are two notions "retirement" in existing code.  There is a temporary retirement that applies only until we rebuild the free set (at end of this gc, or start of next evac).  And there is a permanent retirement that places a fill object into the remnant of free memory so that the memory is no longer free.  (maybe this distinction is not on the master branch.)  The permanent retirement lasts until this region is collected.  The accounting is different.  In both cases, we increase ShenandoahFreeSet::used.  In the permanent case, we also increase the relevant generation:used.

Reply:
I'm not sure that we really need both approaches.  The rationale for permanent retirement is that we don't want to repeatedly discover and re-retire old-gen regions that are scattered throughout the heap.  This is less of a problem with young-gen regions, because regions that are retired once will either soon be promoted or they will be collected.

Reply:
I'm not familiar with the 'permanent' retirement - I don't think that is on `master`. We do fill in LABs with objects when they are retired, but that memory was already 'used' as soon as the LAB was allocated. These changes don't actually touch how the freeset tracks its own usage. What's been hoisted out is only the accounting for generation/heap usage, allocation rate and what the pacer uses. If the 'permanent' retirement includes the filler object in the `actual_size` of the allocation then this new function will account for it as expected. It _should_ do this so that sum of region::used equals the used for a generation/heap.